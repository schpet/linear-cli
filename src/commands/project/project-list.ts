import { Command } from "@cliffy/command";
import { unicodeWidth } from "@std/cli";
import { gql } from "../../__codegen__/gql.ts";
import type {
  GetProjectsQuery,
  ProjectStatusType,
} from "../../__codegen__/graphql.ts";
import { getGraphQLClient } from "../../utils/graphql.ts";
import { getTimeAgo, padDisplay } from "../../utils/display.ts";
import { getTeamKey } from "../../utils/linear.ts";

const GetProjects = gql(`
  query GetProjects($filter: ProjectFilter) {
    projects(filter: $filter) {
      nodes {
        id
        name
        description
        slugId
        icon
        color
        status {
          id
          name
          color
          type
        }
        lead {
          name
          displayName
          initials
        }
        priority
        health
        startDate
        targetDate
        startedAt
        completedAt
        canceledAt
        createdAt
        updatedAt
        url
        teams {
          nodes {
            key
          }
        }
      }
    }
  }
`);

export const listCommand = new Command()
  .name("list")
  .description("List projects")
  .option("--team <team:string>", "Filter by team key")
  .option("--all-teams", "Show projects from all teams")
  .option("--status <status:string>", "Filter by status name")
  .action(async ({ team, allTeams, status }) => {
    const { Spinner } = await import("@std/cli/unstable-spinner");
    const showSpinner = Deno.stdout.isTerminal();
    const spinner = showSpinner ? new Spinner() : null;
    spinner?.start();

    try {
      // Validate conflicting flags
      if (team && allTeams) {
        console.error("Cannot use both --team and --all-teams flags");
        Deno.exit(1);
      }

      // Determine team to filter by
      const teamKey = allTeams ? null : (team?.toUpperCase() || getTeamKey());

      let filter = {};
      if (teamKey) {
        filter = {
          ...filter,
          accessibleTeams: { some: { key: { eq: teamKey } } },
        };
      }
      if (status) {
        filter = { ...filter, status: { name: { eq: status } } };
      }

      const client = getGraphQLClient();
      const result = await client.request(GetProjects, {
        filter: Object.keys(filter).length > 0 ? filter : undefined,
      });
      spinner?.stop();

      let projects = result.projects?.nodes || [];

      if (projects.length === 0) {
        console.log("No projects found.");
        return;
      }

      // Sort projects logically by status then by relevant date
      const statusOrder: Record<ProjectStatusType, number> = {
        "started": 1,
        "planned": 2,
        "backlog": 3,
        "paused": 4,
        "completed": 5,
        "canceled": 6,
      };

      projects = projects.sort((a, b) => {
        // First sort by status type priority
        const statusA =
          statusOrder[a.status.type as keyof typeof statusOrder] || 999;
        const statusB =
          statusOrder[b.status.type as keyof typeof statusOrder] || 999;

        if (statusA !== statusB) {
          return statusA - statusB;
        }

        // Then sort alphabetically by name
        return a.name.localeCompare(b.name);
      });

      // Helper function to get the most relevant date to display
      const getDisplayDate = (
        project: GetProjectsQuery["projects"]["nodes"][0],
      ) => {
        switch (project.status.type) {
          case "started":
            return project.startedAt
              ? `Started ${getTimeAgo(new Date(project.startedAt))}`
              : project.startDate
              ? `Start: ${project.startDate}`
              : `Created ${getTimeAgo(new Date(project.createdAt))}`;
          case "completed":
            return project.completedAt
              ? `Done ${getTimeAgo(new Date(project.completedAt))}`
              : `Updated ${getTimeAgo(new Date(project.updatedAt))}`;
          case "canceled":
            return project.canceledAt
              ? `Canceled ${getTimeAgo(new Date(project.canceledAt))}`
              : `Updated ${getTimeAgo(new Date(project.updatedAt))}`;
          case "planned":
            return project.startDate
              ? `Start: ${project.startDate}`
              : project.targetDate
              ? `Target: ${project.targetDate}`
              : `Created ${getTimeAgo(new Date(project.createdAt))}`;
          case "backlog":
          case "paused":
          default:
            return `Updated ${getTimeAgo(new Date(project.updatedAt))}`;
        }
      };

      // Define column widths based on actual data
      const { columns } = Deno.stdout.isTerminal()
        ? Deno.consoleSize()
        : { columns: 120 };
      const SLUG_WIDTH = Math.max(
        4, // minimum width for "SLUG" header
        ...projects.map((project) => project.slugId.length),
      );
      const STATUS_WIDTH = Math.max(
        6, // minimum width for "STATUS" header
        ...projects.map((project) => project.status.name.length),
      );

      // Calculate priority and health widths based on actual values
      const priorityMap = {
        0: "None",
        1: "Urgent",
        2: "High",
        3: "Medium",
        4: "Low",
      };
      const PRIORITY_WIDTH = Math.max(
        8, // minimum width for "PRIORITY" header
        ...projects.map((project) => {
          const priority =
            priorityMap[project.priority as keyof typeof priorityMap] || "None";
          return priority.length;
        }),
      );
      const HEALTH_WIDTH = Math.max(
        6, // minimum width for "HEALTH" header
        ...projects.map((project) => {
          const health = project.health || "Unknown";
          return health.length;
        }),
      );

      const LEAD_WIDTH = Math.max(
        4, // minimum width for "LEAD" header
        ...projects.map((project) => (project.lead?.initials || "-").length),
      );
      const TEAMS_WIDTH = Math.max(
        5, // minimum width for "TEAMS" header
        ...projects.map((project) => {
          const teams = project.teams.nodes.map((t) => t.key).join(",") || "-";
          return teams.length;
        }),
      );
      const DATE_WIDTH = Math.max(
        4, // minimum width for "DATE" header
        ...projects.map((project) => getDisplayDate(project).length),
      );
      const SPACE_WIDTH = 4;

      const fixed = SLUG_WIDTH + STATUS_WIDTH + PRIORITY_WIDTH + HEALTH_WIDTH +
        LEAD_WIDTH + TEAMS_WIDTH + DATE_WIDTH + SPACE_WIDTH;
      const PADDING = 1;
      const maxNameWidth = Math.max(
        ...projects.map((project) => unicodeWidth(project.name)),
      );
      const availableWidth = Math.max(columns - PADDING - fixed, 0);
      const nameWidth = Math.min(maxNameWidth, availableWidth);

      // Print header
      const headerCells = [
        padDisplay("SLUG", SLUG_WIDTH),
        padDisplay("NAME", nameWidth),
        padDisplay("STATUS", STATUS_WIDTH),
        padDisplay("PRIORITY", PRIORITY_WIDTH),
        padDisplay("HEALTH", HEALTH_WIDTH),
        padDisplay("LEAD", LEAD_WIDTH),
        padDisplay("TEAMS", TEAMS_WIDTH),
        padDisplay("DATE", DATE_WIDTH),
      ];

      let headerMsg = "";
      const headerStyles: string[] = [];
      headerCells.forEach((cell, index) => {
        headerMsg += `%c${cell}`;
        headerStyles.push("text-decoration: underline");
        if (index < headerCells.length - 1) {
          headerMsg += "%c %c";
          headerStyles.push("text-decoration: none");
          headerStyles.push("text-decoration: underline");
        }
      });
      console.log(headerMsg, ...headerStyles);

      // Print each project
      for (const project of projects) {
        const priorityMap = {
          0: "None",
          1: "Urgent",
          2: "High",
          3: "Medium",
          4: "Low",
        };
        const priority =
          priorityMap[project.priority as keyof typeof priorityMap] || "None";
        const health = project.health || "Unknown";
        const lead = project.lead?.initials || "-";
        const teams = project.teams.nodes.map((t) => t.key).join(",") || "-";
        const dateDisplay = getDisplayDate(project);

        const truncName = project.name.length > nameWidth
          ? project.name.slice(0, nameWidth - 3) + "..."
          : padDisplay(project.name, nameWidth);

        console.log(
          `${padDisplay(project.slugId, SLUG_WIDTH)} ${truncName} %c${
            padDisplay(project.status.name, STATUS_WIDTH)
          }%c ${padDisplay(priority, PRIORITY_WIDTH)} ${
            padDisplay(health, HEALTH_WIDTH)
          } ${padDisplay(lead, LEAD_WIDTH)} ${
            padDisplay(teams, TEAMS_WIDTH)
          } %c${padDisplay(dateDisplay, DATE_WIDTH)}%c`,
          `color: ${project.status.color}`,
          "",
          "color: gray",
          "",
        );
      }
    } catch (error) {
      spinner?.stop();
      console.error("Failed to fetch projects:", error);
      Deno.exit(1);
    }
  });
