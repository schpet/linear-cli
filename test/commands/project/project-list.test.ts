import { snapshotTest } from "@cliffy/testing";
import { listCommand } from "../../../src/commands/project/project-list.ts";
import { commonDenoArgs } from "../../utils/test-helpers.ts";
import { MockLinearServer } from "../../utils/mock_linear_server.ts";

// Test help output
await snapshotTest({
  name: "Project List Command - Help Text",
  meta: import.meta,
  colors: true,
  args: ["--help"],
  denoArgs: commonDenoArgs,
  async fn() {
    await listCommand.parse();
  },
});

// Test with mock server - Projects list
await snapshotTest({
  name: "Project List Command - With Mock Projects",
  meta: import.meta,
  colors: false,
  args: ["--all-teams"],
  denoArgs: commonDenoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetProjects",
        variables: { filter: undefined },
        response: {
          data: {
            projects: {
              nodes: [
                {
                  id: "project-1",
                  name: "Authentication System",
                  description: "Core authentication and authorization system",
                  slugId: "auth-sys",
                  icon: "üîê",
                  color: "#3b82f6",
                  status: {
                    id: "status-1",
                    name: "In Progress",
                    color: "#f59e0b",
                    type: "started",
                  },
                  lead: {
                    name: "jane.smith",
                    displayName: "Jane Smith",
                    initials: "JS",
                  },
                  priority: 2,
                  health: "onTrack",
                  startDate: "2024-01-15",
                  targetDate: "2024-03-30",
                  startedAt: "2024-01-16T09:00:00Z",
                  completedAt: null,
                  canceledAt: null,
                  createdAt: "2024-01-10T10:00:00Z",
                  updatedAt: "2024-01-20T15:30:00Z",
                  url: "https://linear.app/test/project/auth-sys",
                  teams: {
                    nodes: [
                      { key: "BACKEND" },
                      { key: "SECURITY" },
                    ],
                  },
                },
                {
                  id: "project-2",
                  name: "Mobile App UI Redesign",
                  description:
                    "Complete redesign of the mobile application interface",
                  slugId: "mobile-ui",
                  icon: "üì±",
                  color: "#ef4444",
                  status: {
                    id: "status-2",
                    name: "Planned",
                    color: "#6366f1",
                    type: "planned",
                  },
                  lead: {
                    name: "alex.designer",
                    displayName: "Alex Designer",
                    initials: "AD",
                  },
                  priority: 3,
                  health: null,
                  startDate: "2024-04-01",
                  targetDate: "2024-06-15",
                  startedAt: null,
                  completedAt: null,
                  canceledAt: null,
                  createdAt: "2024-01-05T14:00:00Z",
                  updatedAt: "2024-01-18T11:15:00Z",
                  url: "https://linear.app/test/project/mobile-ui",
                  teams: {
                    nodes: [
                      { key: "DESIGN" },
                      { key: "MOBILE" },
                    ],
                  },
                },
                {
                  id: "project-3",
                  name: "API Documentation",
                  description: "Comprehensive API documentation and examples",
                  slugId: "api-docs",
                  icon: null,
                  color: "#10b981",
                  status: {
                    id: "status-3",
                    name: "Completed",
                    color: "#059669",
                    type: "completed",
                  },
                  lead: null,
                  priority: 4,
                  health: "onTrack",
                  startDate: "2023-11-01",
                  targetDate: "2024-01-01",
                  startedAt: "2023-11-05T08:00:00Z",
                  completedAt: "2023-12-20T17:30:00Z",
                  canceledAt: null,
                  createdAt: "2023-10-25T09:00:00Z",
                  updatedAt: "2023-12-20T17:30:00Z",
                  url: "https://linear.app/test/project/api-docs",
                  teams: {
                    nodes: [
                      { key: "DOCS" },
                    ],
                  },
                },
              ],
            },
          },
        },
      },
    ]);

    try {
      await server.start();
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint());
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token");

      await listCommand.parse();
    } finally {
      await server.stop();
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT");
      Deno.env.delete("LINEAR_API_KEY");
    }
  },
});

// Test with empty projects list
await snapshotTest({
  name: "Project List Command - No Projects Found",
  meta: import.meta,
  colors: false,
  args: ["--all-teams"],
  denoArgs: commonDenoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetProjects",
        variables: { filter: undefined },
        response: {
          data: {
            projects: {
              nodes: [],
            },
          },
        },
      },
    ]);

    try {
      await server.start();
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint());
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token");

      await listCommand.parse();
    } finally {
      await server.stop();
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT");
      Deno.env.delete("LINEAR_API_KEY");
    }
  },
});
