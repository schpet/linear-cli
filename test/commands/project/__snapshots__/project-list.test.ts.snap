export const snapshot = {};

snapshot[`Project List Command - Help Text 1`] = `
stdout:
"
\\x1b[1mUsage:\\x1b[22m \\x1b[95mlist\\x1b[39m

\\x1b[1mDescription:\\x1b[22m

  List projects

\\x1b[1mOptions:\\x1b[22m

  \\x1b[94m-h\\x1b[39m, \\x1b[94m--help\\x1b[39m             \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Show this help.               
  \\x1b[94m--team\\x1b[39m       \\x1b[33m<\\x1b[39m\\x1b[95mteam\\x1b[39m\\x1b[33m>\\x1b[39m    \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Filter by team key            
  \\x1b[94m--all-teams\\x1b[39m            \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Show projects from all teams  
  \\x1b[94m--status\\x1b[39m     \\x1b[33m<\\x1b[39m\\x1b[95mstatus\\x1b[39m\\x1b[33m>\\x1b[39m  \\x1b[31m\\x1b[1m-\\x1b[22m\\x1b[39m Filter by status name         

"
stderr:
""
`;

snapshot[`Project List Command - With Mock Projects 1`] = `
stdout:
"SLUG      NAME                   STATUS      PRIORITY HEALTH  LEAD TEAMS            DATE                
auth-sys  Authentication System  In Progress High     onTrack JS   BACKEND,SECURITY Started 579 days ago
mobile-ui Mobile App UI Redesign Planned     Medium   Unknown AD   DESIGN,MOBILE    Start: 2024-04-01   
api-docs  API Documentation      Completed   Low      onTrack -    DOCS             Done 605 days ago   
"
stderr:
""
`;

snapshot[`Project List Command - No Projects Found 1`] = `
stdout:
"No projects found.
"
stderr:
""
`;
